### ISSUE TITLE
`git_clone` creates an empty repo if the remote resolves to the same path with the local one

### ISSUE THREAD

**Issue by yuyichao** [libgit2/libgit2#3524]
Ref https://github.com/JuliaLang/julia/issues/14051 and https://github.com/JuliaLang/julia/pull/14146, in particular https://github.com/JuliaLang/julia/pull/14146#issuecomment-159796661 .

When the remote does not exist or is not a git repo, it is expected that `git_clone` should return an error. However, if the remote and the local path happens to resolve to the same path, `git_clone` creates an empty git repo since the local path is initialized before checking the remote. (FWIW the git command line report an error in this case.)

**Comment by tkelman** [comment]
Bump. Has anyone on the libgit2 team looked at this?

> Yep, [`git_clone`](https://github.com/libgit2/libgit2/blob/e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2/src/clone.c#L384) [calles `repository_cb`](https://github.com/libgit2/libgit2/blob/e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2/src/clone.c#L415), which is [`default_repository_create`](https://github.com/libgit2/libgit2/blob/e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2/src/clone.c#L418) by default, [to initialize the local repo](https://github.com/libgit2/libgit2/blob/e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2/src/clone.c#L235) and only [check the remote after that](https://github.com/libgit2/libgit2/blob/e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2/src/clone.c#L423). Therefore, if the url and the local path resolves to the same local directory, it will happily clone the repository it has just initialized and do nothing. This seems like a libgit2 bug for me.

**Comment by ethomson** [comment]
No, this seems rather low priority, and I'm afraid that we have several other things going on at the moment (as well as much reduced bandwidth with the holidays and what not).  A PR would be most welcome if you have the time but this is probably something that we can address eventually if not.

**Comment by tkelman** [comment]
Okay thanks for the response. This is resulting in really strange behavior of cloning something nonexistent and not erroring. Especially noticeable in cases of typos at the REPL. Checks and errors for this condition upstream would seem preferable to workarounds downstream, given a simple acknowledgement of "yes this seems wrong."

**Comment by ethomson** [comment]
I agree - trying to work around this would be quite obnoxious.  It makes sense that libgit2 should fix this however it seems like it would be quite subtle.

I'm brainstorming from my phone so please don't take this as gospel, but the easy seeming solution is to do a comparison of the target / source paths.  But so many things make this subtly ugly:  case sensitivity, relativity, hard links, and weird filesystem equivalences...

A possibility *might* be to try to open the remote before finally creating `HEAD` so that the new repo is not valid until after the target is verified.  Not sure if this is easily doable given the existing code but I suspect it's the most resilient way to solve this problem.  But again I'm mostly just spitballing.

**Comment by tkelman** [comment]
I'll see if I can try some things out. Thanks for the brainstorming, that's helpful for a direction to start looking.
